v2/contribute/code-style/react.mdx

```diff
@@ -6,11 +6,46 @@ hide_title: true
-    - If applicable, make sure to return a function that runs in case the dependency array has changed (or component is unmounting). In this function, you can tell the `useEffect` logic to not change state for example (in case there is an `async / await` operation happening in it.)
+    - If applicable, make sure to return a function that runs in case the dependency array has changed (or the component is unmounting). In this function, you can tell the `useEffect` logic to not change state for example (in case there is an `async / await` operation happening in it.)
 
-- Refreaining returning a component from a function. Instead, make that function itself a functional component. This will prevent that component from unmounting and mounting during rerenders. If you must return a component from a function, then the consumer of that function should use `useRef`.
\ No newline at end of file
+- Refrain from returning a component from a function. Instead, make that function itself a functional component. This will prevent that component from unmounting and mounting during rerenders. If you must return a component from a function, then the consumer of that function should use `useRef` or `useMemo`.
+
+## Feature and Theme components
+
+Complex functionality should be implemented in feature components and theme components should only have local, simple, and strictly UI-related states. Some examples:
+
+### Feature component state:
+- Information that is created in one step of the process and used in another. I.e., `loginAttemptInfo` in passwordless sign in/up
+- General error displayed in the sign-in screen but set in the form submit method
+
+### Theme component state:
+- Current value of each form field
+- Time left until another retry is allowed
+
+### `useReducer`
+
+You should use a reducer to manage the state of a feature component if:
+- The state is complex (i.e.: more than a single error and/or status field)
+- The feature component functionality will be extracted into a custom hook and reused elsewhere.
+- If other similar feature components need to use a reducer. It's better to be consistent. (e.g.: sign in/up feature components of different auth recipes)
+
+If you used a reducer to manage the state you should:
+- Move both state and action types into `types.ts` of the recipe.
+- The "root" theme component should receive them as `featureState` and `dispatch`. Further down the component tree, they are usually event handlers and individual state variables.
+- Check out other components for naming conventions, keeping in mind that it can be renamed if it's imported elsewhere
+
+### Reusable feature components
+
+If the logic of the feature component is reused (i.e.: the corresponding "root" theme component is used in multiple places), it should be extracted into custom hooks. In most cases, these should be `useFeatureReducer` and `useChildProps`.
```

Comment:
If the root theme component is used somewhere else too (like emailpassword's root theme component is used in thirdpartyemailpassword?), then how does adding these custom hook in emailpassword feature component help thirdpartyemailpassword?

Reply:
The main responsibility of the feature component is to provide the business logic behind theme components, they are basically the container component of a single theme component ("root"). 
If we want to reuse the root theme component (or a significant portion of it) we will want to also reuse the business logic implementation, which will be easier if they are implemented in custom hooks. This has a few advantages:
- It's strongly typed: the way we used to do this through cloning the child components and overwriting their props was basically passing anys around.
- It's easier to compose: it's much easier to implement interactions between feature components (more precisely these reused bits of business logic) i.e.: getting a single error variable that we can display instead of both subforms handling/showing their own instance of general error.
- It's a much more standard/recommended way to solve this issue.

Reply:
okay

Reply:
Consider adding this to the guide's MD file too.